### PR TITLE
GH-40896: [Java] Remove runtime dependencies on Eclipse, logback

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

Remove runtime dependencies on [Category B](https://apache.org/legal/resolved.html#category-b) dependencies.

### What changes are included in this PR?

- logback: move to test-only
- eclipse: remove dependency, vendor the Netty implementation we originally used

I wanted to remove javax.annotation.Generated but gRPC doesn't yet let us do that (https://github.com/grpc/grpc-java/issues/9179). That's ~okay though since effectively that's a build only dependency.

### Are these changes tested?

#40901

### Are there any user-facing changes?

No.

**This PR contains a "Critical Fix".** License issues do not cause runtime issues but are important as an Apache project.
* GitHub Issue: #40896

### CODE REVIEW

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Comment by lidavidm** [comment]
CC @danepitkin @vibhatha

**Review by danepitkin — APPROVED** [review]
LGTM. Thanks for the quick fix!

**Comment by danepitkin** [comment]
CI error looks unrelated.

**Comment by lidavidm** [comment]
Oops, I think I need to update either LICENSE.txt or NOTICE to reflect the vendored code too. One sec...

**Review by vibhatha — APPROVED** [review]
LGTM! 
@lidavidm shall we re-run the crossbows just to make sure.

**Comment by lidavidm** [comment]
There is still some time for feature freeze so I'll leave this for a committer to look at. I won't be able to make updates in the next couple days but should be back online by 4/2.

**Comment by assignUser** [comment]
@github-actions crossbow submit *java*

**Review by kou — APPROVED** [review]
+1

File: LICENSE.txt [hunk 1]
```diff
@@ -2252,3 +2252,10 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
+
+--------------------------------------------------------------------------------
+java/vector/src/main/java/org/apache/arrow/vector/util/IntObjectHashMap.java
```
Comment (jbonofre) [inline LICENSE.txt:2257]: Maybe worth to mention in NOTICE file too.

Comment (kou) [reply LICENSE.txt:2257]: If Netty doesn't have `NOTICE` contents for this file, we don't need to change our `NOTICE` file.
See also: https://infra.apache.org/licensing-howto.html#alv2-dep

**Pull request merged**